torch/csrc/jit/codegen/cuda/predicate_compute.cpp

```diff
@@ -106,6 +106,113 @@ std::vector<kir::Bool*> PredicateCompute::computePredicates(
   return preds;
 }
 
+namespace {
+
+//! Analyze whether IterDomain can be statically determined to be safe
+//! without bounds-checking predicates.
+class IterationDomainAnalysis : private IterVisitor {
+ public:
+  //! Return true if the expression defining tv can be safely run
+  //! without a predicate
+  static bool canOmitPredicate(const kir::TensorView* tv) {
+    auto fuser_tv = tv->fuserTv();
+    for (size_t i = 0; i < fuser_tv->nDims(); ++i) {
+      IterDomain* id = fuser_tv->getComputeAtAxis(i).first;
+      auto extent = id->rawExtent();
+      IterationDomainAnalysis id_analysis;
+      id_analysis.traverseFrom(id->fusion(), {extent});
+      if (!id_analysis.isExact(extent)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+ private:
+  using IterVisitor::handle;
+
+  //! Check if the value of val is statically known
+  bool isConst(const Val* val) const {
+    return const_vals_.find(val) != const_vals_.end();
+  }
+
+  //! Return the actual value of val, which must be known to be
+  //! constant
+  int getConst(const Val* val) const {
+    TORCH_INTERNAL_ASSERT(isConst(val));
+    return const_vals_.at(val);
+  }
+
+  //! Record val has a static value of const_val
+  void setConst(const Val* val, int const_val) {
+    TORCH_INTERNAL_ASSERT(const_val > 0);
+    const_vals_[val] = const_val;
+    setExact(val);
+  }
+
+  //! Check if val has nothing that prevents a loop using val as its
+  //! extent to omit a bounds-checking predicate
+  bool isExact(const Val* val) const {
+    return exact_vals_.find(val) != exact_vals_.end();
+  }
+
+  //! Record val does not need a predicate.
+  void setExact(const Val* val) {
+    exact_vals_.insert(val);
+  }
+
+  void handle(Int* val) override {
+    if (val->isConst()) {
+      setConst(val, *(val->value()));
+    } else if (val->definition() == nullptr) {
+      // Undefined symbolic values
+      setExact(val);
+    }
+  }
+
+  void handle(BinaryOp* bop) override {
```

Comment:
Can we use expression evaluation for this or inherit from expression evaluation with the additional handling of ceilDiv?

Reply:
Couldn't we also generate the predicates and use expression evaluation to check that it's always true or always false then simplify?

Reply:
It could inherit from `ExpressionEvaluator`, but it still needs to check `CeilDiv` and `Mul`, so I don't think inheriting would buy us much.

Reply:
> Couldn't we also generate the predicates and use expression evaluation to check that it's always true or always false then simplify?

Predicates are not necessary even for IterDomains with symbolic extents as far as they don't have non-divisible CeilDiv. So, just using `ExpressionEvaluator` isn't sufficient to remove such predicates.

Reply:
I was thinking inheriting from `ExpressionEvaluator` would be a good way to future proof the feature as we may add more operators. Unless we're convinced we'll never use other operators except those enabled here. I don't follow the last point though, could you provide an example where it wouldn't work?

Reply:
I updated it using `ExpressionEvaluator`. It resulted in much simpler code. Thanks for your suggestion.